Netcracker/KubeMarine
# [CPDEV-93684] [CPDEV-93603] [PSUPCLPL-14888] Enrichment in three stages and procedure inventory compilation

### Description
* Support jinja templates in procedure inventory, and in particular env variables.
* Support jinja templates in kubeadm.kubernetesVersion
* Optimize plugins re-installation during upgrade
* Support plugin re-installation during add/remove node if enriched inventory is changed.
    * In particular, re-install `calico` if typha settings are changed or new `route-reflector` is added.

### Solution
* Enrich inventory in three stages:
    * LIGHT (existed) - to fetch information from nodes (e.g. OS family).
    * DEFAULT (new) - to enrich main inventory without applying of procedure inventory.
    * PROCEDURE (existed) - to enrich inventory taking into account procedure inventory.
* Enrich procedure inventory at the earliest point and only then run full enrichment at PROCEDURE stage.
    * This allows to compile procedure inventory - the part that is merged with the main inventory.
      For example, `thirdparties` for `upgrade` procedure, but not `pss.namespaces` for `manage_pss` procedure.
* Compare previous and new inventory after maintenance procedure is applied.
    * Do not re-install OOB plugins during upgrade if effectively resolved inventory does not change (in particular, version compatibility is not changed).
    * Re-install plugins if effectively resolved inventory changes after added/removed nodes (e.g. calico typha settings).
* Other changes:
    * check_iaas handles not accessible nodes and can perform checks partially.
    * check_paas can work on multiple OS family.
    * bugfixes in migrate_kubemarine related to enrichment of few software upgrade patches.

### Potential impact on user experience
* No longer allow to upgrade, if all the following conditions are satisfied:
    * thirdparties are redefined in the main inventory,
    * thirdparties do not depend on kubernetesVersion using jinja templates
    * new thirdparties are not provided in the procedure inventory.
* No longer allow to upgrade if `thirdparties.source` is changed but `thirdparties.sha1` was not.
* No longer allow unrecognized sections in upgrade procedure inventory (e.g. `services.thirdparties` instead of `v1.29.1.thirdparties`)
* Automatic redeploy of plugins in add_node/remove_node procedures if their configurations change.
* Skip redeploy of OOB plugins if version compatibility is not changed. 

### Regression testing
* Literally all test cases for all procedures, and especially:
    * install
    * add/remove node
    * upgrade
    * manage pss/psp
    * cert renew
    * check iaas/paas

### Test Cases

**TestCase 1**

Steps:

1. Supply `services.kubeadm.kubernetesVersion` through env variable and install the cluster.

ER: the cluster is installed successfully.

**TestCase 2**

Steps:

1. Supply `<version>.thirdparties` with jinja templates and install the cluster.

ER: thirdparties are correctly merged with the main inventory and compiled.

**TestCase 3**

Steps:

1. Upgrade v1.27.8 -> v1.28.6 -> v1.29.1 using default inventory.

ER:
1. `calico`, `nginx-ingress-controller`, and `local-path-provisioner` are redeployed for upgrade v1.27.8 -> v1.28.6
2. No plugins are redeployed for upgrade v1.28.6 -> v1.29.1

**TestCase 4**

Steps:

1. Redefine thirdparties in cluster.yaml without use of jinja templates.
2. Run upgrade

ER: The procedure is failed because new thirdparties are not provided.

**TestCase 5**

Steps:

1. Redefine `thirdparties.source` and `thirdparties.sha1` in cluster.yaml.
2. Upgrade 1) without redefining of sha1; 2) with changed source but not changed sha1

ER: The procedure should failed in both cases with different errors.

**TestCase 6**

Steps:

1. Supply procedure inventory with section `services.thirdparties` for upgrade.

ER: The procedure is failed because the section is not expected.

**TestCase 7**

Steps:

1. Install cluster with exactly 3 kubernetes nodes (master, worker, or combined, but not balancer) and 1 dedicated balancer.
   `calico.typha.enabled` should be **unspecified**.

ER: calico-typha is disabled

2. Add 4th kubernetes node.

ER: calico plugin is redeployed with calico-typha enabled.

**TestCase 8**

Steps:

1. Add node with `route-reflector: True` label.
2. Check `calicoctl get nodes -o yaml`

ER: `spec.bgp.routeReflectorClusterID` is patched accordingly.

**TestCase 9**

Steps:

1. Run `check_iaas` with some nodes not accessible.

ER: Checks are failed gracefully, and some checks are still performed.

**TestCase 10**

Steps:

1. Run `check_paas` on cluster with multiple OS families.

ER: Checks are failed gracefully, and most of the checks are still performed.

### Checklist

#### Unit tests
* All tests are adapted for new enrichment
* New tests to cover new and existing functionality in
    * test_run_actions.py, 
    * test_upgrade.py
    * test_migrate_kubemarine.py
    * test_manage_psp.py
    * test_manage.py
    * test_env.py
    * test_admission.py
    * test_k8s_cert.py
    * test_cri_enrichment.py
    * test_calico.py
    * test_flow.py
    * test_cluster.py
